[Exchange] Assertion failed

Sending `0.018188` L-mBTC gives this error:
```
"AssertionError: Assertion failed
    at new push../node_modules/google-protobuf/google-protobuf.js.goog.asserts.AssertionError (http://localhost:8100/static/js/vendors~main.chunk.js:119336:20)
    at Object.push../node_modules/google-protobuf/google-protobuf.js.goog.asserts.doAssertFailure_ (http://localhost:8100/static/js/vendors~main.chunk.js:119365:7)
    at Object.push../node_modules/google-protobuf/google-protobuf.js.goog.asserts.assert (http://localhost:8100/static/js/vendors~main.chunk.js:119374:53)
    at push../node_modules/google-protobuf/google-protobuf.js.jspb.BinaryEncoder.writeUnsignedVarint64 (http://localhost:8100/static/js/vendors~main.chunk.js:124123:16)
    at push../node_modules/google-protobuf/google-protobuf.js.jspb.BinaryWriter.writeUnsignedVarint64_ (http://localhost:8100/static/js/vendors~main.chunk.js:124591:96)
    at push../node_modules/google-protobuf/google-protobuf.js.jspb.BinaryWriter.writeUint64 (http://localhost:8100/static/js/vendors~main.chunk.js:124639:89)
    at Function.push../node_modules/tdex-protobuf/generated/js/trade_pb.js.proto.MarketPriceRequest.serializeBinaryToWriter (http://localhost:8100/static/js/vendors~main.chunk.js:227565:12)
    at push../node_modules/tdex-protobuf/generated/js/trade_pb.js.proto.MarketPriceRequest.serializeBinary (http://localhost:8100/static/js/vendors~main.chunk.js:227536:28)
    at pa.a (http://localhost:8100/static/js/vendors~main.chunk.js:225256:22)
    at Ac (http://localhost:8100/static/js/vendors~main.chunk.js:127863:9)
    at Z.<anonymous> (http://localhost:8100/static/js/vendors~main.chunk.js:127774:12)
    at Z.push../node_modules/grpc-web/index.js.Z.R (http://localhost:8100/static/js/vendors~main.chunk.js:127775:14)
    at TradeClient.marketPrice (http://localhost:8100/static/js/vendors~main.chunk.js:225291:27)
    at http://localhost:8100/static/js/vendors~main.chunk.js:231593:21
    at new Promise (<anonymous>)
    at TraderClient.marketPrice (http://localhost:8100/static/js/vendors~main.chunk.js:231592:12)
    at calculatePrice (http://localhost:8100/static/js/main.chunk.js:27796:33)
    at http://localhost:8100/static/js/main.chunk.js:3066:94"
```

--- a/src/components/ExchangeRow/index.tsx
+++ b/src/components/ExchangeRow/index.tsx
@@ -12,13 +12,12 @@ import { Decimal } from 'decimal.js';
 import { chevronDownOutline } from 'ionicons/icons';
 import { debounce } from 'lodash';
 import React, { useEffect, useMemo, useRef, useState } from 'react';
-import { useSelector } from 'react-redux';
 
 import type { TDEXTrade } from '../../redux/actionTypes/tdexActionTypes';
 import type { BalanceInterface } from '../../redux/actionTypes/walletActionTypes';
 import ExchangeSearch from '../../redux/containers/exchangeSearchContainer';
 import { defaultPrecision } from '../../utils/constants';
-import type { AssetConfig } from '../../utils/constants';
+import type { AssetConfig, LbtcDenomination } from '../../utils/constants';
 import {
   fromSatoshi,
   fromSatoshiFixed,
@@ -66,6 +65,7 @@ interface ExchangeRowInterface {
   setError: (msg: string) => void;
   setOtherInputError: (msg: string) => void;
   isLoading: boolean;
+  lbtcUnit: LbtcDenomination;
 }
 
 const ExchangeRow: React.FC<ExchangeRowInterface> = ({
@@ -90,8 +90,8 @@ const ExchangeRow: React.FC<ExchangeRowInterface> = ({
   setError,
   setOtherInputError,
   isLoading,
+  lbtcUnit,
 }) => {
-  const lbtcUnit = useSelector((state: any) => state.settings.denominationLBTC);
   const [balance, setBalance] = useState<BalanceInterface>();
   const [amount, setAmount] = useState('');
   const [isSearchOpen, setIsSearchOpen] = useState(false);
@@ -244,7 +244,11 @@ const ExchangeRow: React.FC<ExchangeRowInterface> = ({
         return;
       }
       // Sanitize
-      const sanitizedValue = sanitizeInputAmount(e.detail.value, setAmount);
+      const sanitizedValue = sanitizeInputAmount(
+        e.detail.value,
+        setAmount,
+        isLbtc(asset.asset) ? lbtcUnit : undefined,
+      );
       // Set
       setAmount(sanitizedValue);
       onChangeAmount(sanitizedValue);

--- a/src/components/WithdrawRow/index.tsx
+++ b/src/components/WithdrawRow/index.tsx
@@ -18,6 +18,7 @@ import { updatePrices } from '../../redux/actions/ratesActions';
 import {
   fromSatoshi,
   fromSatoshiFixed,
+  isLbtc,
   toLBTCwithUnit,
 } from '../../utils/helpers';
 import { sanitizeInputAmount } from '../../utils/input';
@@ -70,7 +71,7 @@ const WithdrawRow: React.FC<WithdrawRowInterface> = ({
         balance.amount.toString(),
         balance.precision,
         balance.precision,
-        balance.ticker === 'L-BTC' ? lbtcUnit : undefined,
+        isLbtc(balance.asset) ? lbtcUnit : undefined,
       ),
     );
   }, [lbtcUnit, balance.amount]);
@@ -85,7 +86,7 @@ const WithdrawRow: React.FC<WithdrawRowInterface> = ({
         balance.amount.toString(),
         balance.precision,
         balance.precision,
-        balance.ticker === 'L-BTC' ? lbtcUnit : undefined,
+        isLbtc(balance.asset) ? lbtcUnit : undefined,
       ),
     );
     if (price) setFiat('0.00');
@@ -97,14 +98,14 @@ const WithdrawRow: React.FC<WithdrawRowInterface> = ({
       reset();
       return;
     }
-
-    const sanitizedValue = sanitizeInputAmount(e.detail.value, setAmount);
+    const unit = isLbtc(balance.asset) ? lbtcUnit : undefined;
+    const sanitizedValue = sanitizeInputAmount(e.detail.value, setAmount, unit);
     // Set values
     setAmount(sanitizedValue);
     const residualAmount = fromSatoshi(
       balance.amount.toString(),
       balance.precision,
-      balance.ticker === 'L-BTC' ? lbtcUnit : undefined,
+      unit,
     ).sub(sanitizedValue);
     setResidualBalance(
       residualAmount.toNumber().toLocaleString('en-US', {
@@ -114,12 +115,7 @@ const WithdrawRow: React.FC<WithdrawRowInterface> = ({
     );
     if (price) {
       setFiat(
-        toLBTCwithUnit(
-          new Decimal(sanitizedValue),
-          balance.ticker === 'L-BTC' ? lbtcUnit : undefined,
-        )
-          .mul(price)
-          .toFixed(2),
+        toLBTCwithUnit(new Decimal(sanitizedValue), unit).mul(price).toFixed(2),
       );
     }
   };

--- a/src/pages/Exchange/index.tsx
+++ b/src/pages/Exchange/index.tsx
@@ -12,7 +12,6 @@ import classNames from 'classnames';
 import type { UtxoInterface, StateRestorerOpts } from 'ldk';
 import { mnemonicRestorerFromState } from 'ldk';
 import React, { useState, useEffect } from 'react';
-import { useSelector } from 'react-redux';
 import type { RouteComponentProps } from 'react-router';
 import type { Dispatch } from 'redux';
 import { TradeType } from 'tdex-sdk';
@@ -35,7 +34,7 @@ import {
 import { watchTransaction } from '../../redux/actions/transactionsActions';
 import { unlockUtxos } from '../../redux/actions/walletActions';
 import ExchangeRow from '../../redux/containers/exchangeRowContainer';
-import type { AssetConfig } from '../../utils/constants';
+import type { AssetConfig, LbtcDenomination } from '../../utils/constants';
 import {
   defaultPrecision,
   PIN_TIMEOUT_FAILURE,
@@ -63,14 +62,15 @@ import './style.scss';
 const ERROR_LIQUIDITY = 'Not enough liquidity in market';
 
 interface ExchangeProps extends RouteComponentProps {
-  balances: BalanceInterface[];
-  utxos: UtxoInterface[];
-  explorerUrl: string;
-  markets: TDEXMarket[];
-  assets: Record<string, AssetConfig>;
   allAssets: AssetWithTicker[];
+  assets: Record<string, AssetConfig>;
+  balances: BalanceInterface[];
   dispatch: Dispatch;
+  explorerUrl: string;
   lastUsedIndexes: StateRestorerOpts;
+  lbtcUnit: LbtcDenomination;
+  markets: TDEXMarket[];
+  utxos: UtxoInterface[];
 }
 
 const Exchange: React.FC<ExchangeProps> = ({
@@ -83,8 +83,8 @@ const Exchange: React.FC<ExchangeProps> = ({
   allAssets,
   dispatch,
   lastUsedIndexes,
+  lbtcUnit,
 }) => {
-  const lbtcUnit = useSelector((state: any) => state.settings.denominationLBTC);
   const [hasBeenSwapped, setHasBeenSwapped] = useState<boolean>(false);
   // user inputs amount
   const [sentAmount, setSentAmount] = useState<string>();

--- a/src/pages/Operations/index.tsx
+++ b/src/pages/Operations/index.tsx
@@ -28,6 +28,7 @@ import { CurrencyIcon, TxIcon } from '../../components/icons';
 import type { BalanceInterface } from '../../redux/actionTypes/walletActionTypes';
 import WatchersLoader from '../../redux/containers/watchersLoaderContainer';
 import { transactionsByAssetSelector } from '../../redux/reducers/transactionsReducer';
+import type { LbtcDenomination } from '../../utils/constants';
 import {
   defaultPrecision,
   LBTC_TICKER,
@@ -52,7 +53,7 @@ interface OperationsProps extends RouteComponentProps {
   balances: BalanceInterface[];
   prices: Record<string, number>;
   currency: string;
-  lbtcUnit: string;
+  lbtcUnit: LbtcDenomination;
 }
 
 const Operations: React.FC<OperationsProps> = ({

--- a/src/redux/containers/exchangeContainer.tsx
+++ b/src/redux/containers/exchangeContainer.tsx
@@ -8,17 +8,18 @@ import { lastUsedIndexesSelector } from '../selectors/walletSelectors';
 
 const mapStateToProps = (state: any) => {
   return {
+    assets: state.assets,
+    allAssets: allAssets(state),
     balances: balancesSelector(state).filter(
       b =>
         b.amount > 0 &&
         getTradablesAssets(state.tdex.markets, b.asset).length > 0,
     ),
     explorerUrl: state.settings.explorerUrl,
+    lastUsedIndexes: lastUsedIndexesSelector(state),
+    lbtcUnit: state.settings.denominationLBTC,
     markets: state.tdex.markets,
     utxos: allUtxosSelector(state),
-    assets: state.assets,
-    allAssets: allAssets(state),
-    lastUsedIndexes: lastUsedIndexesSelector(state),
   };
 };
 

--- a/src/redux/containers/exchangeRowContainer.tsx
+++ b/src/redux/containers/exchangeRowContainer.tsx
@@ -5,10 +5,11 @@ import { balancesSelector } from '../reducers/walletReducer';
 
 const mapStateToProps = (state: any) => {
   return {
-    prices: state.rates.prices,
-    currency: state.settings.currency.value,
-    balances: balancesSelector(state),
     assets: state.assets,
+    balances: balancesSelector(state),
+    currency: state.settings.currency.value,
+    lbtcUnit: state.settings.denominationLBTC,
+    prices: state.rates.prices,
   };
 };
 

--- a/src/redux/reducers/settingsReducer.ts
+++ b/src/redux/reducers/settingsReducer.ts
@@ -1,3 +1,4 @@
+import type { LbtcDenomination } from '../../utils/constants';
 import { CURRENCIES, LBTC_DENOMINATIONS } from '../../utils/constants';
 import type { ActionType } from '../../utils/types';
 import {
@@ -18,7 +19,7 @@ export interface SettingsState {
   currency: CurrencyInterface;
   explorerUrl: string;
   theme: string;
-  denominationLBTC: string;
+  denominationLBTC: LbtcDenomination;
 }
 
 const initialState: SettingsState = {

--- a/src/utils/constants.ts
+++ b/src/utils/constants.ts
@@ -115,7 +115,13 @@ export const CURRENCIES: CurrencyInterface[] = [
   },
 ];
 
-export const LBTC_DENOMINATIONS = ['L-BTC', 'L-mBTC', 'L-bits', 'L-sats'];
+export const LBTC_DENOMINATIONS = [
+  'L-BTC',
+  'L-mBTC',
+  'L-bits',
+  'L-sats',
+] as const;
+export type LbtcDenomination = typeof LBTC_DENOMINATIONS[number];
 
 export const TOAST_TIMEOUT_SUCCESS = 800;
 export const TOAST_TIMEOUT_FAILURE = 2000;

--- a/src/utils/helpers.ts
+++ b/src/utils/helpers.ts
@@ -19,7 +19,7 @@ import type { BalanceInterface } from '../redux/actionTypes/walletActionTypes';
 import { lockUtxo } from '../redux/actions/walletActions';
 import { network } from '../redux/config';
 
-import type { AssetConfig } from './constants';
+import type { AssetConfig, LbtcDenomination } from './constants';
 import { defaultPrecision, getColor, getMainAsset } from './constants';
 import type { TxDisplayInterface } from './types';
 
@@ -37,23 +37,20 @@ export const createColorFromHash = (id: string): string => {
 
 export function toSatoshi(
   val: string,
-  precision?: number,
-  unit?: string,
+  precision = defaultPrecision,
+  unit: LbtcDenomination = 'L-BTC',
 ): Decimal {
-  const v = new Decimal(val)
-    .mul(Decimal.pow(10, precision ?? defaultPrecision))
-    .floor();
-  return toLBTCwithUnit(v, unit);
+  return new Decimal(val).mul(
+    Decimal.pow(10, new Decimal(precision).minus(unitToExponent(unit))),
+  );
 }
 
 export function fromSatoshi(
   val: string,
-  precision?: number,
-  unit?: string,
+  precision = defaultPrecision,
+  unit: LbtcDenomination = 'L-BTC',
 ): Decimal {
-  const v = new Decimal(val).div(
-    Decimal.pow(10, precision ?? defaultPrecision),
-  );
+  const v = new Decimal(val).div(Decimal.pow(10, precision));
   return formatLBTCwithUnit(v, unit);
 }
 
@@ -61,7 +58,7 @@ export function fromSatoshiFixed(
   val: string,
   precision?: number,
   fixed?: number,
-  unit?: string,
+  unit?: LbtcDenomination,
 ): string {
   return formatLBTCwithUnit(fromSatoshi(val, precision), unit)
     .toNumber()
@@ -72,8 +69,10 @@ export function fromSatoshiFixed(
     });
 }
 
-export function toLBTCwithUnit(lbtcValue: Decimal, unit?: string): Decimal {
-  if (!unit) return lbtcValue;
+export function toLBTCwithUnit(
+  lbtcValue: Decimal,
+  unit?: LbtcDenomination,
+): Decimal {
   switch (unit) {
     case 'L-BTC':
       return lbtcValue;
@@ -88,8 +87,10 @@ export function toLBTCwithUnit(lbtcValue: Decimal, unit?: string): Decimal {
   }
 }
 
-export function formatLBTCwithUnit(lbtcValue: Decimal, unit?: string): Decimal {
-  if (!unit) return lbtcValue;
+export function formatLBTCwithUnit(
+  lbtcValue: Decimal,
+  unit?: LbtcDenomination,
+): Decimal {
   switch (unit) {
     case 'L-BTC':
       return lbtcValue;
@@ -117,6 +118,21 @@ export function unitToFixedDigits(unit?: string): number {
       return 0;
     default:
       return 2;
+  }
+}
+
+export function unitToExponent(unit: LbtcDenomination): number {
+  switch (unit) {
+    case 'L-BTC':
+      return 0;
+    case 'L-mBTC':
+      return 3;
+    case 'L-bits':
+      return 6;
+    case 'L-sats':
+      return 8;
+    default:
+      return 0;
   }
 }
 

--- a/src/utils/input.ts
+++ b/src/utils/input.ts
@@ -1,23 +1,26 @@
 import type { Dispatch } from 'react';
 
-const MAX_DECIMAL_DIGITS = 8;
+import type { LbtcDenomination } from './constants';
+import { unitToFixedDigits } from './helpers';
 
 /**
  * Sanitize input amount
  * @param eventDetailValue string
  * @param setAmount Dispatch<string>
+ * @param unit
  * @returns sanitizedValue string
  */
 export function sanitizeInputAmount(
   eventDetailValue: string,
   setAmount: Dispatch<string>,
+  unit: LbtcDenomination = 'L-BTC',
 ): string {
   // If value is one of those cases, provoke re-rendering with sanitized value
   if (
     // First comma is always replaced by dot. Reset if user types a second comma
     (eventDetailValue.includes('.') && eventDetailValue.includes(',')) ||
-    // No more than MAX_DECIMAL_DIGITS digits
-    eventDetailValue.split(/[,.]/, 2)[1]?.length > MAX_DECIMAL_DIGITS ||
+    // No more than max decimal digits for respective unit
+    eventDetailValue.split(/[,.]/, 2)[1]?.length > unitToFixedDigits(unit) ||
     // If not numbers or dot
     /[^0-9.]/.test(eventDetailValue) ||
     // No more than one dot
@@ -37,9 +40,9 @@ export function sanitizeInputAmount(
   if ((sanitizedValue.match(/\./g) || []).length > 1) {
     sanitizedValue = sanitizedValue.replace(/\.$/, '');
   }
-  // No more than MAX_DECIMAL_DIGITS decimal digits
-  if (eventDetailValue.split(/[,.]/, 2)[1]?.length > MAX_DECIMAL_DIGITS) {
-    sanitizedValue = Number(eventDetailValue).toFixed(MAX_DECIMAL_DIGITS);
+  // No more than max decimal digits for respective unit
+  if (eventDetailValue.split(/[,.]/, 2)[1]?.length > unitToFixedDigits(unit)) {
+    sanitizedValue = Number(eventDetailValue).toFixed(unitToFixedDigits(unit));
   }
   return sanitizedValue === '' ? '0' : sanitizedValue;
 }

--- a/src/utils/tdex.ts
+++ b/src/utils/tdex.ts
@@ -9,6 +9,7 @@ import type {
   TDEXProvider,
 } from '../redux/actionTypes/tdexActionTypes';
 
+import type { LbtcDenomination } from './constants';
 import { getMainAsset } from './constants';
 import { InvalidTradeTypeError, MakeTradeError } from './errors';
 import { isLbtc, toSatoshi } from './helpers';
@@ -29,7 +30,7 @@ export interface AssetWithTicker {
 export async function bestPrice(
   known: { amount: string; asset: string; precision: number },
   trades: TDEXTrade[],
-  lbtcUnit: string,
+  lbtcUnit: LbtcDenomination,
   onError: (e: string) => void,
 ): Promise<{ amount: number; asset: string; trade: TDEXTrade }> {
   if (trades.length === 0) throw new Error('trades array should not be empty');
@@ -88,7 +89,7 @@ export async function bestBalance(trades: TDEXTrade[]): Promise<TDEXTrade> {
 export async function calculatePrice(
   known: { amount: string; asset: string; precision: number },
   trade: TDEXTrade,
-  lbtcUnit: string,
+  lbtcUnit: LbtcDenomination,
 ): Promise<{ amount: number; asset: string }> {
   if (Number(known.amount) <= 0) {
     return {

